app/gui/view/graph-editor/src/component/node/input/widget.rs

```diff
@@ -0,0 +1,163 @@
+//! Definition all variants of possible node widgets and common widget API.
+
+use crate::prelude::*;
+
+use enso_frp as frp;
+use ensogl::application::Application;
+use ensogl::data::color;
+use ensogl::display;
+use ensogl::display::object::event;
+use ensogl_component::drop_down::Dropdown;
+
+
+
+// ==================
+// === NodeWidget ===
+// ==================
+
+ensogl::define_endpoints_2! {
+    Input {
+        set_current_value(Option<ImString>),
+        set_focused(bool),
+    }
+    Output {
+        value_changed(Option<ImString>),
+    }
+}
+
+/// Possible widgets for a node input.
+#[derive(Clone, Debug, CloneRef)]
+pub enum NodeWidget {
```

Comment:
Instead of such Enum, I'd allow any "widget" to be registered on nodes. Just like visualisations. We should allow for custom widgets in the future (as we allow for custom visualisations). We can leave it for now, but in such a case, we should write in the comment that this is a temporary solution that will be removed in <...> task.

Reply:
The design doc lists all the things in order to make it easy to understand. However, the design in the code should be generic, as we know that we will have user-defined widgets sooner than later. Changing such a fundamental assumption later would make a lot of changes in the code, so it's better to do it in the code beforehand. I'd like to make it the extensible way before all the widgets tasks are finished, so just please take care of that when designing the arch here :)